### PR TITLE
Support object count quota for custom resources

### LINKED ISSUE #53777: quota support for custom resources
**Is this a BUG REPORT or FEATURE REQUEST?**:
FEATURE REQUEST

/kind feature

For standard resources like (pods, configmaps), we can use quota to limit the amount of these resources in cluster. However, for a user defined custom resource, we don't support quota.

### LINKED ISSUE #59826: Resource Quota controller should allow discover-resources errors
**Is this a BUG REPORT or FEATURE REQUEST?**:
> /kind bug

**What happened**: On a cluster with an extension API server (service-catalog in this case), Controller Manager crashes with the following log:
controllermanager.go:155] error starting controllers: failed to discover resources: unable to retrieve the complete list of server APIs: servicecatalog.k8s.io/v1beta1: an error on the server ("missing route (no endpoints available for service \"catalog-catalog-apiserver\")") has prevented the request from succeeding

**What you expected to happen**: It should treat these as transient errors and should not return error to the caller (which is main controller-manager). 

**How to reproduce it (as minimally and precisely as possible)**:
Don't really know if it is deterministically reproducible, but a mis-configured (or buggy) extension API server can cause this.

**Anything else we need to know?**:
We ran in to similar issue in 1.7 release and it was fixed in this PR https://github.com/kubernetes/kubernetes/pull/49767/files. The change allowed discovery-resources error in controller-manager. So as long as any resource was returned by discovery API, controller-manager would treat it as transient error.

However in 1.9, one of the controllers (resource-quota) was refactored to accept discoveryFn as input and error handling in discoveryFn doesn't handle the transient errors (which were introduced in 1.7). https://github.com/kubernetes/kubernetes/blob/release-1.9/pkg/controller/resourcequota/resource_quota_controller.go#L468

**Environment**:
- Kubernetes version (use `kubectl version`): 1.9
- Cloud provider or hardware configuration: GKE
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by nikhita**
Fixes  #53777
Fixes #59826
Draws a lot of inspiration from #47665

The resource quota controller now uses a dynamic client and a rest mapper to understand custom resources created via CRDs and aggregated apiservers.

To calculate the quota usage, the quota controller needs informers and listers.

- A new shared informer is created for custom resources.
- The controller also uses a object count quota evaluator to calculate the usage. The evaluator uses a lister to list the resources. So a new lister is created (by creating an indexer) for the custom resource.

Please note that the resource quota controller syncs every 30 seconds, so it might take some time for the quota status to be updated after the CRD is created. Until the quota status is updated and synced, we cannot create any custom resources.

Also, the custom resources created need to be namespaced to work with the quota controller.

**Release note**:

```release-note
Object count quota is now supported for namespaced custom resources using the `count/<resource>.<group>` syntax.
```

### CODE REVIEW

**Comment by nikhita** [comment]
In action with kubectl:

```sh
# create the object count resource quota
$ kubectl create quota test --hard=count/crontabs.example.com=2
resourcequota/test created

# the usage will not be reflected till the CRD is created
$ kubectl get quota test -oyaml
apiVersion: v1
kind: ResourceQuota
metadata:
  creationTimestamp: 2018-05-23T13:39:25Z
  name: test1
  namespace: default
  resourceVersion: "319"
  selfLink: /api/v1/namespaces/default/resourcequotas/test1
  uid: b5826694-5e8e-11e8-8ec1-54e1ad6c2d05
spec:
  hard:
    count/crontabs.example.com: "2"
status:
  hard:
    count/crontabs.example.com: "2"

# create the CRD and wait for sometime (the crd should be available via discovery and the quota controller needs to sync)
$ kubectl create -f crd.yaml
customresourcedefinition.apiextensions.k8s.io/crontabs.example.com created

# the quota now contains the usage
$ kubectl get quota test -oyaml
apiVersion: v1
kind: ResourceQuota
metadata:
  creationTimestamp: 2018-05-23T13:39:37Z
  name: test
  namespace: default
  resourceVersion: "651"
  selfLink: /api/v1/namespaces/default/resourcequotas/test
  uid: bc8d1a09-5e8e-11e8-8ec1-54e1ad6c2d05
spec:
  hard:
    count/crontabs.example.com: "2"
status:
  hard:
    count/crontabs.example.com: "2"
  used:
    count/crontabs.example.com: "0"

# create custom resources
$ kubectl create -f cr1.yaml
crontab.example.com/my-new-cron-object-1 created

$ kubectl create -f cr2.yaml
crontab.example.com/my-new-cron-object-2 created

$ kubectl create -f cr3.yaml
Error from server (Forbidden): error when creating "cr3.yaml": crontabs.example.com "my-new-cron-object-3" is forbidden: exceeded quota: test, requested: count/crontabs.example.com=1, used: count/crontabs.example.com=2, limited: count/crontabs.example.com=2
```

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 1]
```diff
@@ -74,10 +77,19 @@ type ResourceQuotaControllerOptions struct {
...
-	InformerFactory InformerFactory
+	// SharedInformerFactory interfaces with shared informers.
+	SharedInformerFactory informers.SharedInformerFactory
 	// Controls full resync of objects monitored for replenishment.
 	ReplenishmentResyncPeriod controller.ResyncPeriodFunc
+	// dynamicClient is used to work with unknown resources,
+	// like those created via CRDs or aggregated apiservers.
+	DynamicClient dynamic.Interface
+	// RESTMapper can reset itself from discovery.
+	RESTMapper resettableRESTMapper
```
Comment (deads2k) [inline pkg/controller/resourcequota/resource_quota_controller.go]: Is the underlying impl for this threadsafe?  We use a generally dynamic restmapper in the controller context, but I don't know if it is designed to tolerate concurrent resets.

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 2]
```diff
@@ -150,23 +172,21 @@ func NewResourceQuotaController(options *ResourceQuotaControllerOptions) (*Resou
...
 		qm := &QuotaMonitor{
-			informersStarted:  options.InformersStarted,
-			informerFactory:   options.InformerFactory,
-			ignoredResources:  options.IgnoredResourcesFunc(),
-			resourceChanges:   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "resource_quota_controller_resource_changes"),
-			resyncPeriod:      options.ReplenishmentResyncPeriod,
-			replenishmentFunc: rq.replenishQuota,
-			registry:          rq.registry,
+			informersStarted:      options.InformersStarted,
+			ignoredResources:      options.IgnoredResourcesFunc(),
```
Comment (deads2k) [inline pkg/controller/resourcequota/resource_quota_controller.go]: This becomes interesting. 

// TODO we either find a way to discover this or find a way to provide it via config

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 3]
```diff
@@ -436,49 +468,99 @@ func (rq *ResourceQuotaController) Sync(discoveryFunc NamespacedResourcesFunc, p
...
 
 		// Ensure workers are paused to avoid processing events before informers
 		// have resynced.
 		rq.workerLock.Lock()
 		defer rq.workerLock.Unlock()
 
+		// Resetting the REST mapper will also invalidate the underlying discovery
+		// client. This is a leaky abstraction and assumes behavior about the REST
+		// mapper, but we'll deal with it for now.
+		rq.restMapper.Reset()
```
Comment (deads2k) [inline pkg/controller/resourcequota/resource_quota_controller.go]: I thought we passed a restmapper that was already being periodically refreshed.  Why do we need to punch this out?

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 4]
```diff
@@ -416,15 +438,25 @@ func (rq *ResourceQuotaController) replenishQuota(groupResource schema.GroupReso
...
+// the course of detecting new resources.
+func (rq *ResourceQuotaController) Sync(discoveryClient discovery.ServerResourcesInterface, period time.Duration, stopCh <-chan struct{}) {
 	// Something has changed, so track the new state and perform a sync.
 	oldResources := make(map[schema.GroupVersionResource]struct{})
 	wait.Until(func() {
 		// Get the current resource list from discovery.
-		newResources, err := GetQuotableResources(discoveryFunc)
-		if err != nil {
-			utilruntime.HandleError(err)
+		newResources := GetQuotableResources(discoveryClient)
```
Comment (deads2k) [inline pkg/controller/resourcequota/resource_quota_controller.go]: Yeah, it seems like you really want to flush your discovery cache and it doesn't have to do with the restmapper

File: pkg/controller/resourcequota/resource_quota_controller_test.go [hunk 5]
```diff
@@ -78,35 +92,195 @@ func newGenericLister(groupResource schema.GroupResource, items []runtime.Object
...
+type fakeActionHandler struct {
+	// statusCode and content returned by this handler for different method + path.
+	response map[string]FakeResponse
+
+	lock    sync.Mutex
+	actions []fakeAction
+}
+
+// ServeHTTP logs the action that occurred and always returns the associated status code
+func (f *fakeActionHandler) ServeHTTP(response http.ResponseWriter, request *http.Request) {
```
Comment (deads2k) [inline pkg/controller/resourcequota/resource_quota_controller_test.go]: this is a lot of mock I didn't expect to see.

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 3 above)
Comment (sttts) [reply pkg/controller/resourcequota/resource_quota_controller.go]: I think we do, outside of the consuming controllers.

**Comment by lavalamp** [comment]
cc @yliaog

**Comment by lavalamp** [comment]
cc @mbohlool

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 1 above)
Comment (nikhita) [reply pkg/controller/resourcequota/resource_quota_controller.go]: Yes.

https://github.com/kubernetes/kubernetes/blob/28f171bd66937dec8b24a05c4b7a1414432f9fe8/staging/src/k8s.io/client-go/restmapper/discovery.go#L210-L220

File: pkg/controller/resourcequota/resource_quota_controller_test.go [hunk 6]
```diff
@@ -78,35 +92,195 @@ func newGenericLister(groupResource schema.GroupResource, items []runtime.Object
...
+		connection, _, err := hijacker.Hijack()
+		if err != nil {
+			return
+		}
+		defer connection.Close()
+		time.Sleep(30 * time.Second)
+	}
+}
+
+type fakeServerResources struct {
```
Comment (sttts) [inline pkg/controller/resourcequota/resource_quota_controller_test.go]: fakeDiscoveryClient

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 3 above)
Comment (nikhita) [reply pkg/controller/resourcequota/resource_quota_controller.go]: > already being periodically refreshed

The `Sync` is supposed to periodically refresh the restmapper with new discovery information.

The reset occurs within this `Sync` function.

File: pkg/controller/resourcequota/resource_quota_controller_test.go (hunk 5 above)
Comment (sttts) [reply pkg/controller/resourcequota/resource_quota_controller_test.go]: the listWatch test doesn't add anything here. That certainly can go, removing some of the mock code.

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 2 above)
Comment (nikhita) [reply pkg/controller/resourcequota/resource_quota_controller.go]: Done

File: pkg/controller/resourcequota/resource_quota_controller_test.go (hunk 5 above)
Comment (nikhita) [reply pkg/controller/resourcequota/resource_quota_controller_test.go]: Removed the listWatch test, along with some others. Weren't really adding too much. This also removed most of the mock code.

Please see this commit: https://github.com/kubernetes/kubernetes/pull/64201/commits/9247d036326d3f1e17644ce50b2b4c5e65b0101f. If remove those tests seems fine, I'll squash.

File: pkg/controller/resourcequota/resource_quota_controller_test.go (hunk 6 above)
Comment (nikhita) [reply pkg/controller/resourcequota/resource_quota_controller_test.go]: removed the mock code from the quota test but fixed this in the GC test

File: pkg/controller/resourcequota/resource_quota_controller_test.go [hunk 7]
```diff
@@ -452,3 +626,303 @@ func TestAddQuota(t *testing.T) {
...
+		actual := GetQuotableResources(client)
+		if !reflect.DeepEqual(test.quotableResources, actual) {
+			t.Errorf("expected resources:\n%v\ngot:\n%v", test.quotableResources, actual)
+		}
+	}
+}
+
+// TestQuotaControllerSync ensures that a discovery client error
+// will not cause the quota controller to block infinitely.
+
```
Comment (sttts) [inline pkg/controller/resourcequota/resource_quota_controller_test.go]: not: no new line

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 2 above)
Comment (nikhita) [reply pkg/controller/resourcequota/resource_quota_controller.go]: Created https://github.com/kubernetes/kubernetes/pull/64310 to allow it to be provided via config (same as GC)

**Comment by nikhita** [comment]
For the record: we are not going to target this for 1.11. Since this ends up creating a new informer and lister for custom resources, we are increasing the number of times we cache resources (GC does the same as well).

This is solvable if we have something like dynamic informers and listers. We will first focus on creating them and then refocus on the quota and GC controller. See https://github.com/kubernetes/kubernetes/pull/64310#issuecomment-392076699 and https://github.com/kubernetes/kubernetes/issues/64319 for more details.

**Comment by nikhita** [comment]
I'm going to close this PR and reopen a proper fix once @p0lyn0mial's PR (https://github.com/kubernetes/kubernetes/pull/69308) gets in.

**Comment by p0lyn0mial** [comment]
> I'm going to close this PR and reopen a proper fix once @p0lyn0mial's PR (#69308) gets in.

@nikhita done :)

**Comment by raushan2016** [comment]
Is this feature now supported. Where i want to limit number of instance of a CRD created in a namespace like number of knative services created in a namespace ?

**Comment by liggitt** [comment]
Support is merged in master and will be part of 1.15

**Comment by raushan2016** [comment]
@liggitt  can you point me to sample/doc about how to use it for CRD objects

**Comment by liggitt** [comment]
the docs at https://kubernetes.io/docs/concepts/policy/resource-quotas/#object-count-quota are accurate

in 1.15+, to quota a custom resource "widgets" in the "example.com" API group, you would use `count/widgets.example.com`

**Comment by liggitt** [comment]
opened doc PR for the 1.15 branch in https://github.com/kubernetes/website/pull/14492

**Pull request closed without merging**